jax/_src/compiler.py

```diff
@@ -314,23 +315,69 @@ def compile_or_get_cached(
     # storage.
     # TODO: In case when coordinator process is not participating in the
     # computation we need to choose another host to compile the module.
-    if distributed.global_state.service is None:
-      serialized_executable = global_client.blocking_key_value_get_bytes(
-          module_name, share_timeout)
-      serialized_executable = compilation_cache.decompress_executable(
-          serialized_executable)
-      executable = backend.deserialize_executable(
-          serialized_executable, compile_options)
-    else:
+    if distributed.global_state.process_id == 0:
       executable = backend_compile(backend, computation,
                                    compile_options, host_callbacks)
       serialized_executable = backend.serialize_executable(executable)
       serialized_executable = compilation_cache.compress_executable(
           serialized_executable)
       global_client.key_value_set(module_name, serialized_executable)
-
+    else:
+      serialized_executable = global_client.blocking_key_value_get_bytes(
+          module_name, share_timeout)
+      serialized_executable = compilation_cache.decompress_executable(
+          serialized_executable)
+      executable = backend.deserialize_executable(
+          serialized_executable, compile_options)
     compile_or_get_cached.cached_modules[module_name] = executable
     return executable
+  elif (
+      process_count() > 1
+      and config.share_autotune_config_between_hosts.value
+      and distributed.global_state.client is not None
+  ):
+    share_timeout = config.share_binary_between_hosts_timeout_ms.value
+    global_client = distributed.global_state.client
+    debug_options = compile_options.executable_build_options.debug_options
+    autotune_tmp_file = os.path.join(
+        compile_or_get_cached.autotune_configs_dir, module_name
+    )
+
+    # If module were previously compiled use existing config.
+    if os.path.exists(autotune_tmp_file):
+      debug_options.xla_gpu_load_autotune_results_from = autotune_tmp_file
+      executable = backend_compile(
+          backend, computation, compile_options, host_callbacks
+      )
+      return executable
+
+    # The coordinator process should compile the module and write autotune
+    # config to to the K-V storage.
+    # TODO: In case when coordinator process is not participating in the
+    # computation we need to choose another host to compile the module.
+    if distributed.global_state.process_id == 0:
+      debug_options.xla_gpu_dump_autotune_results_to = autotune_tmp_file
+      executable = backend_compile(
```

Comment:
Could we stop the compilation after all the autotuning?
This would help lower the extra compilation time.
Do you have any estimate how much extra compilation time is done right now? 2x? Or significantly lower due to the saving in autotuner time?

Reply:
To generate autotune we might need to run a whole optimization passes IIUC. We still can save codegen part, but it's usually takes a few percents of compilation time.
I checked the following task on PAXML (https://github.com/google/paxml/blob/main/paxml/tasks/lm/params/nvidia.py#L402) but on 2 hosts and here what I observed:

1. Compiling with autotune config provided is significantly faster then without it. So for example on one run the compilation time for jit_psum module on the first host (the one which dumps autotune config) ~6.8 seconds vs 0.54 on the second host.
An overall compilation time was 6.86 s on the host which dumps autotune config vs 10.24s on the second host (waiting for the config from the first host ~9.7s + compiling the module again 0.5s.)

2. The module with the same name might need to be recompiled. For example we might need to recompile jit_reshape with different paramenters. So we can't use only module name as sync key (fixed at the last commit).

Are other multi-host tasks which you think might worth to check?

Reply:
From your number, the compilation is 1.5x slower and 1.68x slower with sharing the autotuner results. The transfer time for 2 nodes(I suppose of 8 GPUs, 1 process per GPUs, so 16 process) was: ~0.5s and ~1s.
Having any idea this scale would be good. We need that to scale to more nodes. Could you try with more nodes and get a sense of the scaling? The issue is probably related to the transfer.

Why do you say the code gen part is fast? In the past, I saw it being the bottleneck. The call to ptxas was costly. Do you include it in your "codegen" timing? If it is still true that ptxas take a significant amount of time, it could be good to find a way to skip it.

Reply:
These are nodes with 16 GPUs (32 processes total). I'll try to collect this information for 4 nodes (64 processes total).
I'll also try to measure codegen timing.

Reply:
I checked a task on 6 nodes (96 GPUs) and numbers are pretty much the same:

P_sum compilation time:
Coordinator host: 20.75s
Other hosts: ~28.5s (28s waiting for autotune config, 0.5s compilation)

jit_convert_element_type compilation time:
Coordinator host: 17.7s
Other hosts: 18.5s (~18s waiting time, 0.5s compilation)

jit__threefry_seed compilation time:
Coordinator host: 11s
Other hosts: 11.3s (~11s waiting time, 0.3s compilation)

jit_concatenate compilation time:
Coordinator host: 16s
Other hosts: 16.5s (~16s waiting time, 0.5s compilation)

jit_reshape
Coordinator host: 12.4s
Other hosts: 18.5s (~18.4s waiting time, 0.1s compilation)

pjit_init_model_from_seed
Coordinator host: 27.3s
Other hosts: 37.9s (~27s waiting time, ~11s compilation)

pjit__wrapped_step_fn:
Coordinator host: 437.4s
Other hosts: 565s (~438s waiting time, ~127.3s compilation)


etc.

Reply:
Regarding codegen, actually at this stage we are only getting an optimized HLO, not the ptx. So the only thing we can eliminate is buffer assignment IIUC.